### PR TITLE
Use an auth helper that can handle digest authentication #284

### PR DESCRIPTION
**Pull request by ryanjdew**
#284

### CODE REVIEW

**Comment by grtjn** [comment]
Looks like some good improvements to me, though maybe someone better at nodejs could have a look.

Just wondering how this compares to #220. I know that has a different aim, but it splits auth functionality into a separate module as well. Could they be combined, or piled up on top of each other?

**Comment by joemfb** [comment]
I'd also like to see this based on `passport`, as that's a widely-used and well known approach

**Comment by grtjn** [comment]
Ryan, could you rebase this PR? Afraid there have been quite a few changes in the middle-tier lately, so might take a bit of effort.

Once you have it running again, can you see if you can look at the PR from hunterwilliams, and merge in use of passport into this PR?

**Comment by grtjn** [comment]
Let me know if you are short on time, or could use some help from anyone..

**Comment by ryanjdew** [comment]
Rebased off master and currently working on getting the passport PR integrated.

**Comment by ryanjdew** [comment]
The passport changes from Hunter have been merged and I tested with basic, digest, and digestbasic authentication settings on the MarkLogic HTTP server.

File: app/templates/node-server/routes.js [hunk 1]
```diff
@@ -3,155 +3,93 @@
...
-  var password = req.body.password || '';
-  var headers = req.headers;
-
-  //make sure login isn't cached
-  noCache(res);
-
-  var startsWithMatch = new RegExp('^' + options.appName + '-');
-  if (options.appUsersOnly && !startsWithMatch.test(username)) {
-    res.status(403).send('Forbidden');
+  if (!req.isAuthenticated()) {
```
Comment (grtjn) [inline app/templates/node-server/routes.js]: I'm missing guestAccess stuff. Can passport authentication work with defaultUser/defaultPass?

File: app/templates/node-server/routes.js [hunk 2]
```diff
@@ -3,155 +3,93 @@
...
+    res.send({authenticated: false});
```
Comment (grtjn) [inline app/templates/node-server/routes.js]: Can you use authStatus for all auth status responses?

**Comment by grtjn** [comment]
The passport PR (from hunter) seems considerably shorter. Is the added complexity in this PR for handling digest authentication?

**Comment by ryanjdew** [comment]
Yes. There is added complexity in that the digest Authorization header has to be calculated and returned prior to making calls to MarkLogic.

**Comment by ryanjdew** [comment]
Added back in the logic for setting options for guest access, app-only users, etc.

**Comment by grtjn** [comment]
Now that I have CI in place, I'll just merge, and check outcome on CI server..

**Comment by grtjn** [comment]
This also fixes #175..

**Pull request merged**